### PR TITLE
kubeadm config add support for more than one APIEndpoint

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR completes the changes in kubeadm for management of more than one control plane instances introducing the possibility to configure more than one APIEndpoints

**Which issue(s) this PR fixes** :
refs https://github.com/kubernetes/kubeadm/issues/911, refs https://github.com/kubernetes/kubeadm/issues/963

**Special notes for your reviewer**:
Depends on:
- [x] https://github.com/kubernetes/kubernetes/pull/67830

**Release note**:
```release-note
kubeadm: The kubeadm configuration now support definition of more than one control plane instances with their own APIEndpoint. The APIEndpoint for the "bootstrap" control plane instance should be defined using `InitConfiguration.APIEndpoint`, while the APIEndpoints for additional control plane instances should be added using `JoinConfiguration.APIEndpoint`.  
```

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/sig cluster-lifecycle
/area kubeadm
/kind api-change
/kind enhancement
/assign @luxas
/assign @timothysc
/cc @chuckha @rosti @neolit123 @liztio

### CODE REVIEW

**Comment by fabriziopandini** [comment]
/hold
until https://github.com/kubernetes/kubernetes/pull/67830 merge

File: cmd/kubeadm/app/apis/kubeadm/types.go
Comment (timothysc) [inline cmd/kubeadm/app/apis/kubeadm/types.go]: The Naming is confusing here. 

The APIEndpoint struct has an APIEndpoint member.

**Review by timothysc — CHANGES_REQUESTED** [review]
This doesn't look right.  

Also perhaps I'm missing the [] or map

File: cmd/kubeadm/app/apis/kubeadm/types.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: @timothysc  Sorry but I don't understand you comment :confused:

APIEndpoint struct has two attributes, AdvertiseAddress and BindPort:

Here we have an attribute that has the same name of its type. is that the thing you find confusing? Do you prefer naming the struct `APIEndpointConfiguration`

Comment (timothysc) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: Sorry, the way github collapsed this changeset made me misinterpret this.  I expanded it looks ok.

Comment (fabriziopandini) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: Or... are you looking for the "status" we discusse at the end of the call yesterday?

If this is the case, please consider that this PR implements only the "spec", while the status will be implemented the next PR (in the form of `map[bindaddress]APIEndoint`, but I'm figuring out details right now)

Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: > map[bindaddress]APIEndoint

i though more about this. we ideally need to map against the address the user provides instead of always defaulting to the address of the network interface, because we do have `--apiserver-advertise-address`.

if i understand this correctly this might result in an interesting situation, but i guess we can comment more in the next PR. :\

File: cmd/kubeadm/app/cmd/phases/certs.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/certs.go]: @kad ^ FYI

File: cmd/kubeadm/app/util/config/masterconfig.go
Comment (neolit123) [inline cmd/kubeadm/app/util/config/masterconfig.go]: just a nit: some of these empty new lines in the function can be removed.
please, live as is for now if you'd prefer.

File: cmd/kubeadm/app/util/config/masterconfig.go
Comment (neolit123) [inline cmd/kubeadm/app/util/config/masterconfig.go]: this will probably conflict with:
https://github.com/kubernetes/kubernetes/pull/67397
either this or that PR will need a rebase.

**Review by neolit123 — COMMENTED** [review]
thanks a lot @fabriziopandini 
added some small comments LGTM!

File: cmd/kubeadm/app/cmd/phases/certs.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/certs.go]: @kad @neolit123 this PR doesn't implemented/changed this bit, only refactored the API struct.
If something should be fixed, this should be address in a separated PR (this one is already XXL 😉)

File: cmd/kubeadm/app/apis/kubeadm/v1alpha2/types.go
Comment (timothysc) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha2/types.go]: Shouldn't this be omitempty, b/c previous versions of v1alpha2 didn't have this field.

File: cmd/kubeadm/app/apis/kubeadm/v1alpha3/types.go
Comment (timothysc) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha3/types.go]: This doesn't appear correct `json:"featureGates,`

**Review by timothysc — APPROVED** [review]
looking good, but I think some the meta-data in your structs needs cleanup, otherwise LGTM.  Please clean that up and we're good to go. 

/approve

**Review by rosti — APPROVED** [review]
Looks OK to me. A few places were converted back to InitConfiguration (from ClusterConfiguration), but I assume, that this way the change is narrower and PR is much more easy to review.

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve 

Thx @fabriziopandini

**Pull request merged**
